boa/src/builtins/json/mod.rs

```diff
@@ -145,133 +148,538 @@ impl Json {
         args: &[JsValue],
         context: &mut Context,
     ) -> Result<JsValue> {
-        let object = match args.get(0) {
-            None => return Ok(JsValue::undefined()),
-            Some(obj) => obj,
-        };
-        const SPACE_INDENT: &str = "          ";
-        let gap = if let Some(space) = args.get(2) {
-            let space = if let Some(space_obj) = space.as_object() {
-                if let Some(space) = space_obj.borrow().as_number() {
-                    JsValue::new(space)
-                } else if let Some(space) = space_obj.borrow().as_string() {
-                    JsValue::new(space)
-                } else {
-                    space.clone()
-                }
-            } else {
-                space.clone()
-            };
-            if space.is_number() {
-                let space_mv = match space.to_integer_or_infinity(context)? {
-                    IntegerOrInfinity::NegativeInfinity => 0,
-                    IntegerOrInfinity::PositiveInfinity => 10,
-                    IntegerOrInfinity::Integer(i) if i < 1 => 0,
-                    IntegerOrInfinity::Integer(i) => std::cmp::min(i, 10) as usize,
-                };
-                JsValue::new(&SPACE_INDENT[..space_mv])
-            } else if let Some(string) = space.as_string() {
-                JsValue::new(&string[..std::cmp::min(string.len(), 10)])
+        // 1. Let stack be a new empty List.
+        let stack = Vec::new();
+
+        // 2. Let indent be the empty String.
+        let indent = JsString::new("");
+
+        // 3. Let PropertyList and ReplacerFunction be undefined.
+        let mut property_list = None;
+        let mut replacer_function = None;
+
+        let replacer = args.get(1).cloned().unwrap_or_default();
+
+        // 4. If Type(replacer) is Object, then
+        if let Some(replacer_obj) = replacer.as_object() {
+            // a. If IsCallable(replacer) is true, then
+            if replacer_obj.is_callable() {
+                // i. Set ReplacerFunction to replacer.
+                replacer_function = Some(replacer_obj)
+            // b. Else,
             } else {
-                JsValue::new("")
+                // i. Let isArray be ? IsArray(replacer).
+                // ii. If isArray is true, then
+                if replacer_obj.is_array() {
+                    // 1. Set PropertyList to a new empty List.
+                    let mut property_set = HashSet::new();
+
+                    // 2. Let len be ? LengthOfArrayLike(replacer).
+                    let len = replacer_obj.length_of_array_like(context)?;
+
+                    // 3. Let k be 0.
+                    let mut k = 0;
+
+                    // 4. Repeat, while k < len,
+                    while k < len {
+                        // a. Let prop be ! ToString(𝔽(k)).
+                        let prop = k.to_string();
```

Comment:
```suggestion
                        let prop = k;
```
we don't have to call `to_string` on k since this would just cause an unnecessary string allocation,  we have `PropertyKey::index` we can just pass an integer to `.get`

Reply:
Done :)